src/RadixDlt.NetworkGateway.Abstractions/CoreCommunications/TransactionSubmitter.cs

```diff
@@ -0,0 +1,240 @@
+        TransactionApi transactionApi,
+        string networkName,
+        byte[] notarizedTransaction,
+        SubmitContext submitContext,
+        IReadOnlyCollection<ITransactionSubmitterObserver> observers,
+        TimeSpan submissionTimeout,
+        CancellationToken cancellationToken)
+    {
+        await observers.ForEachAsync(observer => observer.ObserveSubmitAttempt(submitContext));
+
+        using var ctsWithSubmissionTimeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        ctsWithSubmissionTimeout.CancelAfter(submissionTimeout);
+
+        var result = await SubmitInternal(transactionApi, networkName, notarizedTransaction, ctsWithSubmissionTimeout.Token);
+
+        await observers.ForEachAsync(observer => observer.ObserveSubmitResult(submitContext, result));
+
+        return result;
+    }
+
+    private static async Task<NodeSubmissionResult> SubmitInternal(
+        TransactionApi transactionApi,
+        string networkName,
+        byte[] notarizedTransaction,
+        CancellationToken cancellationToken)
+    {
+        try
+        {
+            var result = await CoreApiErrorWrapper.ResultOrError<TransactionSubmitResponse, TransactionSubmitErrorResponse>(() =>
+                transactionApi.TransactionSubmitPostAsync(
+                    new TransactionSubmitRequest(
+                        network: networkName,
+                        notarizedTransactionHex: notarizedTransaction.ToHex()
+                    ),
+                    cancellationToken
+                ));
+
+            if (result.Succeeded)
+            {
+                return new NodeSubmissionResult.AcceptedIntoMempool(IsAlreadyThere: result.SuccessResponse.Duplicate);
+            }
+
+            return result.FailureResponse.Details switch
+            {
+                TransactionSubmitIntentAlreadyCommitted alreadyCommittedDetails =>
+                    new NodeSubmissionResult.IntentAlreadyCommitted(alreadyCommittedDetails.CommittedAs),
+                TransactionSubmitPriorityThresholdNotMetErrorDetails priorityThresholdDetails =>
+                    new NodeSubmissionResult.MempoolPriorityThresholdNotMet(priorityThresholdDetails),
+                TransactionSubmitRejectedErrorDetails rejectedDetails =>
+                    new NodeSubmissionResult.Rejection(rejectedDetails),
+                _ => throw new ArgumentOutOfRangeException(nameof(result.FailureResponse.Details)),
+            };
+        }
+        catch (OperationCanceledException ex)
+        {
+            return new NodeSubmissionResult.OtherSubmissionError(ex, "Operation timed-out");
+        }
+        catch (Exception ex)
+        {
+            return new NodeSubmissionResult.OtherSubmissionError(ex, "Unknown error");
+        }
+    }
+}
+
+public abstract record NodeSubmissionResult
+{
+    // ======================================================================
+    // WARNING WHEN ADDING VARIANTS:
```

Comment:
While Rust and its pattern matching is no doubt a great technique, C# isn't designed this way. IMO we should avoid those constructs if possible. Otherwise this codebase is going to become super error-prone maintenance nightmare.

For instance `ShouldStopSubmittingPermanently` should be simply overridden by implementations to return their specific logic; we don't need type-based pattern matching here. Same's true for intent and payload status.

Reply:
👍 true - better to make these abstract methods

Reply:
That said, I am a big fan of allowing pattern matching, and capturing all possible outcomes into such a sum type. It provides better encapsulation of the concept of a result, and you're less likely to miss things. (I accept that we should avoid switches ofc, your suggestion to use abstract methods is better).

EG see how nicely abstracted the metrics are now, and it makes it much easier to handle all cases, instead of missing eg an Exception.